pygmt/datasets/tile_map.py

```diff
@@ -128,6 +138,8 @@ def load_tile_map(
     ...     raster.rio.crs.to_string()
     'EPSG:3857'
     """
+    _default_crs = "EPSG:3857"  # The default CRS returned from contextily
```

Comment:
At https://contextily.readthedocs.io/en/latest/reference.html#contextily.bounds2img, there is this sentence:

> IMPORTANT: tiles are assumed to be in the Spherical Mercator projection (EPSG:3857), unless the crs keyword is specified.

While tilemaps are usually served in EPSG:3857, I know that there are some tilemaps that can be served in a different projection system by default. I'm hesitant to use this default of EPSG:3857 here, unless we can be confident that contextily only supports EPSG:3857 tiles.

Reply:
Looking at the contexily source codes, I think contexitly always assumes that CRS is EPSG:3857, at least when writing the image to raster files

https://github.com/geopandas/contextily/blob/eafeb3c993ab43356b65e7a29b9dd6df3ac76f00/contextily/tile.py#L358

Reply:
That is for the `warp_tiles` function. I don't see anything about EPSG:3857 in the [`bounds2img`](https://github.com/geopandas/contextily/blob/00ca0318033e69e29a164ce571d9c932a057ecc6/contextily/tile.py#L183-L293) function (though it's calling several functions and I haven't checked too closely).

Non-Web Mercator (EPSG:3857) XYZ tiles are not common, but they do exist, e.g. Openlayers allows configuring the projection system of the returned tiles:
- https://www.linz.govt.nz/guidance/data-service/linz-data-service-guide/map-tile-services/using-lds-xyz-services-openlayers#c-0-s-4
- https://openlayers.org/en/main/apidoc/module-ol_source_XYZ-XYZ.html
- https://stackoverflow.com/questions/65024840/xyz-in-non-web-mercator-projections

Reply:
> Looking at the contexily source codes, I think contexitly always assumes that CRS is EPSG:3857, at least when writing the image to raster files

I meant this line:

https://github.com/geopandas/contextily/blob/eafeb3c993ab43356b65e7a29b9dd6df3ac76f00/contextily/tile.py#L175

In bounds2raster, it first calls `bounds2img` to get the `image` and `extent` (can be any CRS), but when writing to a raster image, `bounds2raster` assumes the `image` is `EPSG:3857`.

Looking at the xyzservices (https://github.com/geopandas/xyzservices/blob/c847f312d2e67a0a45ceae119fdcdf7cf60858b6/provider_sources/xyzservices-providers.json#L50), I can also see some providers that has the `crs` attribute, but it's unclear how the `crs` attribute is handled in contexitly.

Reply:
I tried using `xyzservices.providers.MapTiler.Basic4326` which comes in EPSG:4326:

```python
import contextily
import xyzservices
import matplotlib.pyplot as plt

source = xyzservices.providers.MapTiler.Basic4326(
    key="BwTZdryxoypHc4BnCZ4"  # add an 'o' to the end
)
img, bounds = contextily.bounds2img(w=0, s=-80, e=150, n=80, ll=True, source=source)
# img.shape # (4096, 2048, 4)
plt.imshow(img[:, :, :3])
```

produces

![temp](https://github.com/user-attachments/assets/18590066-0e51-4c2e-b4c4-2b10a18fcd2f)

The extent doesn't seem to be correct (it should be plotting from Greenwich to Asia), possibly a bug in `contextily` hardcoding to EPSG:3857. But it does show that non-Web Mercator XYZ tiles are indeed possible.

Reply:
Found https://github.com/geopandas/contextily/issues/119, and had a closer look at the `bounds2img` code, specifically these lines - https://github.com/geopandas/contextily/blob/00ca0318033e69e29a164ce571d9c932a057ecc6/contextily/tile.py#L270-L272. It seems like `contextily` is using [`mercantile`](https://github.com/mapbox/mercantile) which only supports Spherical Mercator (EPSG:3857), see https://github.com/mapbox/mercantile/issues/109#issuecomment-636331797. Unless `contextily` switches to something like https://github.com/developmentseed/morecantile in the future, then it would only 'work' with EPSG:3857.

Reply:
> The extent doesn't seem to be correct (it should be plotting from Greenwich to Asia), possibly a bug in `contextily` hardcoding to EPSG:3857.

https://github.com/geopandas/contextily/blob/eafeb3c993ab43356b65e7a29b9dd6df3ac76f00/contextily/tile.py#L256

When `ll=True`, `bounds2img` calls the private [`_sm2ll`](https://github.com/geopandas/contextily/blob/eafeb3c993ab43356b65e7a29b9dd6df3ac76f00/contextily/tile.py#L539) function to convert bounds in Spherical Mercator coordinates to lon/lat. I think the contextily package assumes "Spherical Mercator projection" in many places.

Reply:
How about checking `source.crs` (an attribute from an instance of the `xyzservices.TileProvider` class), and if it is present, use that as the default crs. If not, fallback to EPSG:3857.

Reply:
Sounds good. Done in 6733e196a6ec5b845e235cbe9f44581b94d2c10d.

> `_default_crs = getattr(source, "crs", "EPSG:3857")`

Please note that this line of code works for `TileProvide`, `str` and `None`. `str` and `None` don't have `crs`, so `EPSG:3857` is returned.